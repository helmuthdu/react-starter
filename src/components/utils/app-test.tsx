import { RouterContext } from 'next/dist/shared/lib/router-context';
import { IntlProvider } from 'react-intl';
import { RecoilRoot } from 'recoil';

const router = jest.mock('next/router', () => {
  const mockedRouter = {
    back: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
    route: '/mock-route',
    pathname: 'mock-path'
  };
  return {
    default: mockedRouter,
    useRouter: jest.fn(() => mockedRouter),
    withRouter: (component: any) => {
      component.defaultProps = {
        ...component.defaultProps,
        router: mockedRouter
      };
      return component;
    }
  };
});

type Props = { children: any };
export const AppTest = ({ children }: Props) => (
  <RouterContext.Provider value={router as any}>
    <RecoilRoot>
      <IntlProvider locale="en" onError={() => undefined}>
        {children}
      </IntlProvider>
    </RecoilRoot>
  </RouterContext.Provider>
);
