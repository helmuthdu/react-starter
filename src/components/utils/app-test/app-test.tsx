import React from 'react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import { store } from '../../../stores';

export const AppTest = ({ children }: any) => (
  <IntlProvider locale="en" onError={jest.fn()}>
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  </IntlProvider>
);
