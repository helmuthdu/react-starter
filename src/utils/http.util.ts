import fetch from 'isomorphic-unfetch';
import { Logger } from './logger.util';

type HttpRequestConfig = RequestInit & { id?: string };

const log = (type: keyof typeof Logger, url: string, req: RequestInit, res: unknown, time: number) => {
  const _url = url?.split('/') as string[];
  const timestamp = Logger.getTimestamp();
  Logger.groupCollapsed(
    `Http.${req.method?.toLowerCase()}('â€¦/${_url[_url.length - 1]}')`,
    `HTTP|${type.toUpperCase()}`,
    time
  );
  Logger.setTimestamp(false);
  Logger.info('url:', url);
  Logger.debug('req:', req);
  Logger[type]('res:' as never, res);
  Logger.setTimestamp(timestamp);
  Logger.groupEnd();
};

const activeRequests = {} as Record<string, { request: Promise<any>; controller: AbortController }>;

export class Http {
  private static _headers: Record<string, string> = {
    'Content-Type': 'application/json'
  };

  static async get<T>(url: string, config?: HttpRequestConfig, latestOnly?: boolean): Promise<T> {
    return await this._request<T>(url, { method: 'GET', ...config }, latestOnly);
  }

  static async post<T>(url: string, config?: HttpRequestConfig, latestOnly?: boolean): Promise<T> {
    return await this._request<T>(url, { method: 'POST', ...config }, latestOnly);
  }

  static async put<T>(url: string, config?: HttpRequestConfig, latestOnly?: boolean): Promise<T> {
    return await this._request<T>(url, { method: 'PUT', ...config }, latestOnly);
  }

  static async patch<T>(url: string, config?: HttpRequestConfig, latestOnly?: boolean): Promise<T> {
    return await this._request<T>(url, { method: 'PATCH', ...config }, latestOnly);
  }

  static async delete<T>(url: string, config?: HttpRequestConfig, latestOnly?: boolean): Promise<T> {
    return await this._request<T>(url, { method: 'DELETE', ...config }, latestOnly);
  }

  public static setHeaders(headers: Record<string, string | undefined>) {
    Object.entries(headers).forEach(([key, val]) => {
      if (val === undefined) {
        delete this._headers[key];
      } else {
        this._headers[key] = val;
      }
    });
  }

  private static async _request<T>(url: string, config: HttpRequestConfig, latestOnly?: boolean): Promise<T> {
    const { id = this.generateRequestId(config), ...cfg } = config;

    if (activeRequests[id] && latestOnly) {
      activeRequests[id].controller.abort();
    }

    if (!activeRequests[id]) {
      const controller = new AbortController();
      const signal = controller.signal;
      const request = this._makeRequest(id, url, { ...cfg, signal });
      activeRequests[id] = { request, controller };
    }

    return activeRequests[id].request;
  }

  private static _makeRequest<T>(id: string, url: string, options: RequestInit): Promise<T> {
    const { headers, ...rest } = options;

    const req: RequestInit = {
      headers: {
        ...this._headers,
        ...headers
      },
      ...rest
    };

    const time = Date.now();
    return fetch(url, req)
      .then(async (res: Response) => {
        const data: T = await res.json();
        log('success', url, req, data, time);
        return data;
      })
      .catch(error => {
        log('error', url, req, error, time);
        throw error;
      })
      .finally(() => {
        delete activeRequests[id];
      });
  }

  private static generateRequestId(options: any): string {
    return `${JSON.stringify(options)}`;
  }
}
