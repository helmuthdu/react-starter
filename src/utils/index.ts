export * from './list.utils';
export * from './search.util';
