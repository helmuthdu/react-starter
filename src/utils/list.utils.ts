/* eslint-disable */
// https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore

export const compose = <R>(fn: (args: R) => R, ...fns: ((args: R) => R)[]) =>
  fns.reduce((prevFn, nextFn) => value => prevFn(nextFn(value)), fn);

export const first = <T>(list: T[], total = 1): T[] => list.slice(0, total);

export const flatten = (list: unknown | unknown[]): unknown[] =>
  Array.isArray(list) ? list.reduce((acc, cur) => acc.concat(flatten(cur)), []) : list;

export const get = <T, K extends keyof T>(obj: T, path: K | string, defaultValue: unknown = null) =>
  String.prototype.split
    .call(path, /[,[\].]+?/)
    .filter(Boolean)
    .reduce((acc: any, cur: string) => (Object.hasOwnProperty.call(acc, cur) ? acc[cur] : defaultValue), obj);

export const groupBy = <T>(list: T | T[] | ReadonlyArray<T>, key: keyof T): DictionaryArray<T> =>
  Object.values(list).reduce(
    (acc: DictionaryArray<T>, val: T, idx: number, arr: T[] | ReadonlyArray<T>, prop = val[key]) => {
      if (!prop) return acc;
      if (!acc[prop]) acc[prop] = [];
      acc[prop].push(val);
      return acc;
    },
    {}
  );

export const keyBy = <T>(list: T | T[] | ReadonlyArray<T>, key: keyof T): Dictionary<T> =>
  Object.values(list).reduce(
    (acc: Dictionary<T>, val: T, idx: number, arr: T[] | ReadonlyArray<T>, prop = val[key]) => {
      if (!prop) return acc;
      acc[prop] = val;
      return acc;
    },
    {}
  );

export const keys = <T, K extends keyof T>(obj: T) => Object.keys(obj) as K[];

export const values = <T, K extends keyof T>(obj: T) => Object.values(obj) as T[K][];

export const last = <T>(arr: T[], total = 1): T[] => arr.slice(-total);

export const pipe = <T extends unknown[], R>(fn: (...args: T) => R, ...fns: ((args: R) => R)[]) => (...args: T) =>
  fns.reduce(
    (prevFn, nextFn) => (value: R) => nextFn(prevFn(value)),
    value => value
  )(fn(...args));

export const sortBy = <T, K extends keyof T>(arr: T[], key: K) =>
  [...arr].sort((a: T, b: T) => (a[key] > b[key] ? 1 : b[key] > a[key] ? -1 : 0));

export const uniq = (arr: number[]) => [...new Set(arr)];

export const type = (val: any) => {
  const typeOf = typeof val;

  if (val === null) {
    return 'Null';
  } else if (val === undefined) {
    return 'Undefined';
  } else if (typeOf === 'boolean') {
    return 'Boolean';
  } else if (typeOf === 'number') {
    return Number.isNaN(val) ? 'NaN' : 'Number';
  } else if (typeOf === 'string') {
    return 'String';
  } else if (Array.isArray(val)) {
    return 'Array';
  } else if (val instanceof RegExp) {
    return 'RegExp';
  }

  const asStr = val.toString();

  if (asStr.startsWith('async')) {
    return 'Async';
  } else if (asStr === '[object Promise]') {
    return 'Promise';
  } else if (typeOf === 'function') {
    return 'Function';
  }

  return 'Object';
};

export const isEquals = function (a: any, b: any) {
  if (arguments.length === 1) return (_b: any) => isEquals(a, _b);

  if (a === b) {
    return true;
  }

  const aType = type(a);

  if (aType !== type(b)) {
    return false;
  }

  if (aType === 'Array') {
    const aClone = Array.from(a);
    const bClone = Array.from(b);

    if (aClone.toString() !== bClone.toString()) {
      return false;
    }

    let loopArrayFlag = true;
    aClone.forEach((aCloneInstance, aCloneIndex) => {
      if (loopArrayFlag) {
        if (aCloneInstance !== bClone[aCloneIndex] && !isEquals(aCloneInstance, bClone[aCloneIndex])) {
          loopArrayFlag = false;
        }
      }
    });

    return loopArrayFlag;
  }

  if (aType === 'Object') {
    const aKeys = Object.keys(a);

    if (aKeys.length !== Object.keys(b).length) {
      return false;
    }

    let loopObjectFlag = true;
    aKeys.forEach(aKeyInstance => {
      if (loopObjectFlag) {
        const aValue = a[aKeyInstance];
        const bValue = b[aKeyInstance];

        if (aValue !== bValue && !isEquals(aValue, bValue)) {
          loopObjectFlag = false;
        }
      }
    });

    return loopObjectFlag;
  }

  return false;
};

export const isEmpty = (val: any): boolean =>
  [Object, Array].includes((val || {}).constructor) && !Object.entries(val || {}).length;

export const isArray = (arr: any) => arr && [Array].includes((arr || {}).constructor);

export const isObject = (obj: any) => obj && [Object].includes((obj || {}).constructor);

export const toSnakeCase = (str: string) =>
  str
    .replace(/\W+/g, ' ')
    .split(/ |\B(?=[A-Z])/)
    .map(s => s.toLowerCase())
    .join('_');

export const toKebabCase = (str: string) =>
  str
    ? str
        .replace(/([a-z])([A-Z])/g, '$1-$2') // get all lowercase letters that are near to uppercase ones
        .replace(/[\s_]+/g, '-') // replace all spaces and low dash
        .toLowerCase() // convert to lower case
    : '';
