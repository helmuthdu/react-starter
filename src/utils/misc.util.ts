/* eslint-disable */
// https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore

export const get = <T, K extends keyof T>(obj: T, path: K | string, defaultValue: unknown = null) =>
  String.prototype.split
    .call(path, /[,[\].]+?/)
    .filter(Boolean)
    .reduce((acc: any, cur: string) => (Object.hasOwnProperty.call(acc, cur) ? acc[cur] : defaultValue), obj);

export const groupBy = <T>(list: T | T[] | ReadonlyArray<T>, key: keyof T) =>
  Object.values(list).reduce(
    (acc: DictionaryArray<T>, val: T, idx: number, arr: T[] | ReadonlyArray<T>, prop = val[key]) =>
      (acc[prop] || (acc[prop] = [])).push(val),
    {}
  );

export const sortBy = <T, K extends keyof T>(arr: T[], key: K) =>
  [...arr].sort((a: T, b: T) => (a[key] > b[key] ? 1 : b[key] > a[key] ? -1 : 0));

export const keyBy = <T>(list: T | T[] | ReadonlyArray<T>, key: keyof T): Dictionary<T> =>
  Object.values(list).reduce(
    (acc: Dictionary<T>, val: T, idx: number, arr: T[] | ReadonlyArray<T>, prop = val[key]) => {
      if (!prop) return acc;
      acc[prop] = val;
      return acc;
    },
    {}
  );

export const uniq = (arr: number[]) => [...new Set(arr)];
export const flatten = <T>(list: T | T[]) => (Array.isArray(list) ? list.flat(Infinity) : list);

export const keys = <T, K extends keyof T>(obj: T) => Object.keys(obj) as K[];
export const values = <T, K extends keyof T>(obj: T) => Object.values(obj) as T[K][];

export const first = <T>(list: T[], total = 1): T[] => list.slice(0, total);
export const last = <T>(arr: T[], total = 1): T[] => arr.slice(-total);

export const compose = <R>(fn: (args: R) => R, ...fns: ((args: R) => R)[]) =>
  fns.reduce((prevFn, nextFn) => value => prevFn(nextFn(value)), fn);
export const pipe =
  <T extends unknown[], R>(fn: (...args: T) => R, ...fns: ((args: R) => R)[]) =>
  (...args: T) =>
    fns.reduce(
      (prevFn, nextFn) => (value: R) => nextFn(prevFn(value)),
      value => value
    )(fn(...args));

type MergeOutput<T> = Required<{ [K in keyof T]: T[K] }>;
export const merge = <T extends Record<string, any>>(target: T, ...sources: T[]): MergeOutput<T> => {
  const source = sources.shift();
  if (!source) return target as MergeOutput<T>;
  Object.keys(source).forEach(key => {
    if (isObject(source[key])) {
      if (!target[key]) Object.assign(target, { [key]: {} });
      merge(target[key], source[key]);
    } else {
      Object.assign(target, { [key]: source[key] });
    }
  });
  return merge(target, ...sources);
};

export const type = (val: any) => {
  if (val === null) {
    return 'Null';
  } else if (val === undefined) {
    return 'Undefined';
  } else if (Number.isNaN(val)) {
    return 'NaN';
  }
  const cast = Object.prototype.toString.call(val).slice(8, -1);
  return cast === 'AsyncFunction' ? 'Promise' : cast;
};

export const isEquals = (a: any, b: any): boolean => {
  if (a === b) return true;

  if (type(a) !== type(b)) return false;

  if (isArray(type(a))) {
    const _a = Array.from(a);
    const _b = Array.from(b);
    if (_a.toString() !== _b.toString()) return false;
    return !_a.some((val, idx) => val !== _b[idx] && !isEquals(val, _b[idx]));
  }

  if (isObject(type(a))) {
    const keys = Object.keys(a);
    if (keys.length !== Object.keys(b).length) return false;
    return !keys.some(key => a[key] !== b[key] && !isEquals(a[key], b[key]));
  }

  return false;
};

export const isArray = (val: unknown) => type(val) === 'Array';
export const isFunction = (val: unknown) => type(val) === 'Function';
export const isNumber = (val: unknown) => type(val) === 'Number';
export const isObject = (val: unknown) => type(val) === 'Object';
export const isString = (val: unknown) => type(val) === 'String';
export const isEmpty = (val: any) =>
  [Object, Array].includes((val || {}).constructor) && !Object.entries(val || {}).length;

export const toSnakeCase = (str: string) =>
  str
    .replace(/\W+/g, ' ')
    .split(/ |\B(?=[A-Z])/)
    .map(s => s.toLowerCase())
    .join('_');

export const toKebabCase = (str: string) =>
  str
    .replace(/([a-z])([A-Z])/g, '$1-$2')
    .replace(/[\s_]+/g, '-')
    .toLowerCase();

export const uuid = (): string => window.crypto.getRandomValues(new Uint32Array(1))[0].toString(16);
