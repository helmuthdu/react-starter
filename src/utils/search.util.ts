import { Observable, SchedulerLike, Subject } from 'rxjs';
import { debounceTime, distinctUntilChanged, filter } from 'rxjs/operators';

interface InputObservableOptions {
  time?: number;
  scheduler?: SchedulerLike;
  minLength?: number;
}

/**
 * Creates an observable variable to be used with a search input
 * @param subject
 * @param options
 * @return Observable<string>
 */
export const searchDebounceObserver = (
  subject: Subject<string>,
  options: InputObservableOptions = {}
): Observable<string> => {
  const { time = 400, scheduler, minLength = 3 } = options;
  return subject.pipe(
    debounceTime(time, scheduler),
    filter(query => query.length >= minLength || query.length === 0),
    distinctUntilChanged()
  );
};
