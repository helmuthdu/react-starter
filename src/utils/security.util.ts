export const generateUniqueId = (): string => window.crypto.getRandomValues(new Uint32Array(1))[0].toString(16);
