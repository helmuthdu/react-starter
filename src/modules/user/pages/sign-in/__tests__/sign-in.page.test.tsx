import { AppTest } from '@/components/utils/app-test';
import { render } from '@testing-library/react';
import SignInPage from '../sign-in.page';

describe('user/page -> SignIn', () => {
  const props: any = {};

  it('should match snapshot', () => {
    const { asFragment } = render(
      <AppTest>
        <SignInPage {...props} />
      </AppTest>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
