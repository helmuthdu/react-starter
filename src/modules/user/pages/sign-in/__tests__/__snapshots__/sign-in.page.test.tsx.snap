// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user/page -> SignIn > should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <input
      name="email"
      type="email"
    />
    <input
      name="password"
      type="password"
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
  User: 
</DocumentFragment>
`;

exports[`user/page -> SignIn should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <input
      name="email"
      type="email"
    />
    <input
      name="password"
      type="password"
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
  User: 
</DocumentFragment>
`;
